internal/attestation/azure/snp/validator.go

```diff
@@ -36,249 +41,368 @@ type Validator struct {
-		validateCVM,
+		// stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
+		func(vtpm.AttestationDocument, *attest.MachineState) error {
+			return nil
+		},
 		log,
 	)
 	return v
 }
 
-// validateCVM is a stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
-func validateCVM(vtpm.AttestationDocument, *attest.MachineState) error {
-	return nil
-}
-
-func newSNPReportFromBytes(reportRaw []byte) (snpAttestationReport, error) {
-	var report snpAttestationReport
-	if err := binary.Read(bytes.NewReader(reportRaw), binary.LittleEndian, &report); err != nil {
-		return snpAttestationReport{}, fmt.Errorf("reading attestation report: %w", err)
-	}
-
-	return report, nil
-}
-
-func reverseEndian(b []byte) {
-	for i := 0; i < len(b)/2; i++ {
-		b[i], b[len(b)-i-1] = b[len(b)-i-1], b[i]
-	}
-}
-
 // getTrustedKey establishes trust in the given public key.
-// It does so by verifying the SNP attestation statement in instanceInfo.
+// It does so by verifying the SNP attestation document.
 func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, extraData []byte) (crypto.PublicKey, error) {
+	// transform the instanceInfo received from Microsoft into a verifiable attestation report format.
 	var instanceInfo azureInstanceInfo
 	if err := json.Unmarshal(attDoc.InstanceInfo, &instanceInfo); err != nil {
-		return nil, fmt.Errorf("unmarshalling instanceInfoRaw: %w", err)
+		return nil, fmt.Errorf("unmarshalling instanceInfo: %w", err)
 	}
-
-	report, err := newSNPReportFromBytes(instanceInfo.AttestationReport)
+	att, err := instanceInfo.attestationWithCerts(v.log, v.getter)
 	if err != nil {
 		return nil, fmt.Errorf("parsing attestation report: %w", err)
 	}
 
-	vcek, err := v.validateVCEK(instanceInfo.Vcek, instanceInfo.CertChain)
+	// Verify the attestation report's certificates.
+	trustedArk := x509.Certificate(v.config.AMDRootKey)             // ARK, specified in Constellation config.
+	ask, err := x509.ParseCertificate(att.CertificateChain.AskCert) // ASK, as reported from THIM / KDS.
 	if err != nil {
-		return nil, fmt.Errorf("validating VCEK: %w", err)
-	}
-
-	if err := v.validateSNPReport(ctx, vcek, report, instanceInfo.MAAToken, extraData); err != nil {
-		return nil, fmt.Errorf("validating SNP report: %w", err)
-	}
-
+		return nil, fmt.Errorf("parsing ASK certificate: %w", err)
+	}
+	// We verify the ASK certificate against the ARK certificate, since the ARK certificate is the root certificate
+	// and inherently trusted. This check should also be performed by go-sev-guest, but as we have to specify both ASK and ARK for
+	// the trusted roots, this needs to be done here as well.
+	rootPool := x509.NewCertPool()
+	rootPool.AddCert(&trustedArk)
+	if _, err := ask.Verify(x509.VerifyOptions{
+		Roots: rootPool,
+	}); err != nil {
+		return nil, fmt.Errorf("verifying ASK certificate: %w", err)
+	}
+	if err := v.attestationVerifier.SnpAttestation(att, &verify.Options{
```

Comment:
To me it seems that the library does it correctly here
https://github.com/google/go-sev-guest/blob/81d183f8e73dee86e453bcfacb454af3d3880abb/verify/verify.go#L376C1-L376C66
because of Options created as
https://github.com/google/go-sev-guest/blob/81d183f8e73dee86e453bcfacb454af3d3880abb/verify/trust/trust.go#L235-L239
Can you add a test with (self-generated) ASK not matching the ARK? It should fail with the verification here. And if you then remove the verification, it should fail in the library.

Reply:
Re "Adding test":
yes I will add this

Why we need this check at all:
The library unfortunately doesn't allow to "only" specify the ARK (root cert) for establishing trust in the chain. It always requires ASK and ARK to be present, since AMD also publishes them together publicly [here](https://github.com/google/go-sev-guest/blob/bab68c55ee83798dc313c70cf9d98b61055e6cb9/verify/trust/trust.go#L37-L39). In the Constellation config, the user is only able to specify ARK. Therefore, to be able to use this ARK at all, we have to verify that it signs the ASK (which the library does as well, as you said) to be able to not use the embedded certificates from the library and let the user explicitly specify which root cert he wants to trust. We have multiple possible solutions to this:
- Do it as we do, verify that the ARK from the Constellation config signs the received ASK (from THIM or KDS) first, then pass this ARK-ASK-pair (ARK from Constellation config, ASK from THIM or KDS) into the library which checks them against the ARK-ASK-pair (both from THIM or KDS) 
- Put the whole CertChain in the Constellation config, this would probably be most consistent with what AMD "expects you to do", since they publish the whole certchain publicly as well, and not just the ARK, as mentioned above.
- Remove the ARK field from the Constellation config and embed the root certs ourselves in the Constellation CLI. We will have to issue a security fix either way in case of a disaster, therefore I don't see any particular reason to have the field in the config except for transparency reasons.
- **Preferred solution:** Get something upstreamed in the library that will allow you to only specify the ARK for the trusted roots. /etc/profiles/per-user/moritzs/bin/go test -timeout 30s -run ^TestTrustedKeyFromSNP$ github.com/edgelesssys/constellation/v2/internal/attestation/azure/snp would be the relevant code part presumably.

Reply:
cc @derpsteb @katexochen

Reply:
Ok with me to pass ARK and ASK. It's just about whether it's required to verify the ASK against the ARK if the library does it as well.

Reply:
I assume passing the ASK directly would cause the trust to only be established from there (i.e. ARK wouldn't be respected).  I will inspect the relevant code parts and add the test you've mentioned to clarify this

Reply:
Indeed, the library does not seem to error on an invalid ASK if we remove this check and just pass it into the trusted roots.